### PR TITLE
WIP make Uvicorn work with wsproto 0.13

### PR DESCRIPTION
**Pull request by almarklein**
This is a work in progress for making things work with wsproto 0.13. For reference, here's [wsprotos changelog](https://github.com/python-hyper/wsproto/blob/master/CHANGELOG.rst#0130-2019-01-24).

I started out assuming that it would need just some renaming in a few places, but it turned out to involve a bit more work. Not sure how to best proceed with this, e.g. trying to make both 0.12 and 0.13 work from one class, create an additional class, or maybe drop 0.12 support?

The tests currently pass (on my machine, with wsproto 0.13) but someone with more experience with this part of the codebase should do a thorough review, or (probably better) pick this up from here.

References for moving from 0.12 to 0.13:
* https://gitlab.com/pgjones/hypercorn/commit/5b635eba3914445e242457bba93e41e940aec34b
* https://github.com/python-hyper/wsproto/blob/master/CHANGELOG.rst#0130-2019-01-24

### CODE REVIEW

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: This corresponds to "Switch from Fail events being returned to RemoteProtocolError`s being raised." in the changelog.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: Not sure where the extensions should gol the new class does not have an argument/attribute for it.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: All these bits correspond to: "Switch from a bytes_to_send method to the send method returning the bytes to send directly. "

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: `target` and `path` are now a string. I *think* `extra_headers` is what previously was `headers`.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py:195]: Cannot send below stuff during the handshake. I suspect this `if self.conn.connection is None` check should be done in a few other places (maybe where `_upgrade_connection` was previously used?).

**Comment by lovelydinosaur** [comment]
> maybe drop 0.12 support

Yup, that'd be the thing to do.

**Comment by lovelydinosaur** [comment]
Thanks for taking a crack at this - let me know if you need extra eyes on it.

**Comment by almarklein** [comment]
> > maybe drop 0.12 support
>
> Yup, that'd be the thing to do.

What about users who need wsproto 0.12 somwhere else in their code? Or can we assume that wsproto is an implementation detail of Uvicorn so unlikely to happen?

**Comment by lovelydinosaur** [comment]
> What about users who need wsproto 0.12 somwhere else in their code? Or can we assume that wsproto is an implementation detail of Uvicorn so unlikely to happen?

We can't reasonably be expected to maintain cross-version compatability if there are API changes in our dependencies. If someone *needs* wsproto 0.12 within the same codebase as uvicorn, then they can also stick with uvicorn 0.4.3 or below.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [reply uvicorn/protocols/websockets/wsproto_impl.py]: These are now passed in the `AcceptConnection` event.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [reply uvicorn/protocols/websockets/wsproto_impl.py]: Confirmed.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py:149]: It seems like this `handle_no_connect()` method is not covered by the test suite ...

**Comment by almarklein** [comment]
Ok, I made some more updates, and also took a look at Hypercorn's implementation for reference. The tests pass, but please have a thorough look at this. I also left some `todo`'s which should probably be resolved before merging.

I'll leave it to you to either review and request changes, or to take this over from here.

File: tests/protocols/test_http.py
Comment (lovelydinosaur) [inline tests/protocols/test_http.py:698]: What's this switch for?

File: tests/protocols/test_websocket.py
Comment (lovelydinosaur) [inline tests/protocols/test_websocket.py:83]: Same question

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (lovelydinosaur) [inline uvicorn/protocols/websockets/wsproto_impl.py]: Okay, that's helpful to know for review, but we should drop the comment now.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (lovelydinosaur) [inline uvicorn/protocols/websockets/wsproto_impl.py]: Yes looks like it to me. Had a quick look at the `wsproto` docs, and extra_headers + host seems to make sense here.

**Review by lovelydinosaur — CHANGES_REQUESTED** [review]

File: tests/protocols/test_http.py
Comment (almarklein) [reply tests/protocols/test_http.py:698]: It's related to the `RemoteProtocolError ` and its `event_hint` value: it responds with 426 (Upgrade Required). Unfortunately, there is no body.

File: tests/protocols/test_websocket.py
Comment (almarklein) [reply tests/protocols/test_websocket.py:83]: Same answer :)

File: tests/protocols/test_http.py
Comment (lovelydinosaur) [reply tests/protocols/test_http.py:698]: > Unfortunately, there is no body.

Just to double check: Do you mean there's no status phrase? (Eg. `HTTP/1.1 426 Upgrade Required`) or there's no body?

Comment (almarklein) [reply tests/protocols/test_http.py:698]: Either, I think.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: @tomchristie what do you think?

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (almarklein) [inline uvicorn/protocols/websockets/wsproto_impl.py]: @tomchristie any comment on this? Maybe something to look at in another PR.

Comment (lovelydinosaur) [reply uvicorn/protocols/websockets/wsproto_impl.py]: Yes, but I'd suggest we treat that as a separate issue.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (lovelydinosaur) [reply uvicorn/protocols/websockets/wsproto_impl.py]: Let's add that and drop the comment.

File: uvicorn/protocols/websockets/wsproto_impl.py
Comment (lovelydinosaur) [reply uvicorn/protocols/websockets/wsproto_impl.py]: We probably don't need it there. `wsproto` ought to take take of giving us a consistent stream of messages. I'd drop the `break` assuming the tests still pass.

**Comment by lovelydinosaur** [comment]
Looks good - just needs the various points you commented on cleaning up now.

**Comment by almarklein** [comment]
I addressed the last points and cleaned up a bit.

**Review by lovelydinosaur — APPROVED** [review]

**Comment by lovelydinosaur** [comment]
Thanks so much for this!

**Pull request merged**
